### PR TITLE
KAFKA-2476: Add Decimal, Date, and Timestamp logical types.

### PR DESCRIPTION
**Pull request by ewencp**
To support Decimal, this also adds support for schema parameters, which is an
extra set of String key value pairs which provide extra information about the
schema. For Decimal, this is used to encode the scale parameter, which is part
of the schema instead of being passed with every value.

### CODE REVIEW

File: copycat/api/src/test/java/org/apache/kafka/copycat/data/CopycatSchemaTest.java [hunk 1]
```diff
@@ -208,14 +208,15 @@ public void testValidateValueMismatchStructWrongNestedSchema() {
...
-        CopycatSchema s1 = new CopycatSchema(Schema.Type.INT8, false, null, "name", 2, "doc", null, null, null);
-        CopycatSchema s2 = new CopycatSchema(Schema.Type.INT8, false, null, "name", 2, "doc", null, null, null);
-        CopycatSchema differentType = new CopycatSchema(Schema.Type.INT16, false, null, "name", 2, "doc", null, null, null);
-        CopycatSchema differentOptional = new CopycatSchema(Schema.Type.INT8, true, null, "name", 2, "doc", null, null, null);
-        CopycatSchema differentDefault = new CopycatSchema(Schema.Type.INT8, false, true, "name", 2, "doc", null, null, null);
-        CopycatSchema differentName = new CopycatSchema(Schema.Type.INT8, false, null, "otherName", 2, "doc", null, null, null);
-        CopycatSchema differentVersion = new CopycatSchema(Schema.Type.INT8, false, null, "name", 4, "doc", null, null, null);
-        CopycatSchema differentDoc = new CopycatSchema(Schema.Type.INT8, false, null, "name", 2, "other doc", null, null, null);
+        CopycatSchema s1 = new CopycatSchema(Schema.Type.INT8, false, null, "name", 2, "doc", null, null, null, null);
```
Comment (guozhangwang) [inline copycat/api/src/test/java/org/apache/kafka/copycat/data/CopycatSchemaTest.java]: Could we have additional CopycatSchema constructors that put default field values as null?

File: copycat/api/src/test/java/org/apache/kafka/copycat/data/SchemaBuilderTest.java [hunk 2]
```diff
@@ -188,6 +190,21 @@ public void testBytesBuilderInvalidDefault() {
...
+
+        Schema schema = SchemaBuilder.string().parameter("foo", "val").parameter("bar", "baz").build();
+        assertTypeAndDefault(schema, Schema.Type.STRING, false, null);
+        assertMetadata(schema, null, null, null, expectedParameters);
+
+        schema = SchemaBuilder.string().parameters(expectedParameters).build();
+        assertTypeAndDefault(schema, Schema.Type.STRING, false, null);
+        assertMetadata(schema, null, null, null, expectedParameters);
+    }
+
```
Comment (guozhangwang) [inline copycat/api/src/test/java/org/apache/kafka/copycat/data/SchemaBuilderTest.java:207]: Add another test case for invalidating parameter values as non-strings?

File: copycat/json/src/main/java/org/apache/kafka/copycat/json/JsonConverter.java [hunk 3]
```diff
@@ -443,6 +527,12 @@ private static JsonNode convertToJson(Schema schema, Object value) {
             throw new DataException("Conversion error: null value for field that is required and has no default value");
         }
 
+        if (schema != null && schema.name() != null) {
+            LogicalTypeConverter logicalConverter = TO_JSON_LOGICAL_CONVERTERS.get(schema.name());
+            if (logicalConverter != null)
+                value = logicalConverter.convert(schema, value);
```
Comment (guozhangwang) [inline copycat/json/src/main/java/org/apache/kafka/copycat/json/JsonConverter.java]: Do we want to overwrite the passed-in value reference here?

File: copycat/api/src/test/java/org/apache/kafka/copycat/data/CopycatSchemaTest.java (hunk 1 above)
Comment (ewencp) [inline copycat/api/src/test/java/org/apache/kafka/copycat/data/CopycatSchemaTest.java]: We definitely could. I expected that people would almost always use `SchemaBuilder` to construct `Schema` objects so I wasn't that worried about making a variety of constructors available. Aside from tests, we only directly construct these objects in a one or two locations.

**Comment by guozhangwang** [comment]
LGTM except minor comments. Could you rebase?

File: copycat/api/src/test/java/org/apache/kafka/copycat/data/SchemaBuilderTest.java (hunk 2 above)
Comment (ewencp) [inline copycat/api/src/test/java/org/apache/kafka/copycat/data/SchemaBuilderTest.java:207]: Not sure I understand. The only two ways to pass in parameters are via `SchemaBuilder parameter(String propertyName, String propertyValue)` and `SchemaBuilder parameters(Map<String, String> props)`. Doesn't that already ensure non-string values won't be passed in?

**Comment by ewencp** [comment]
@guozhangwang merged and addressed most of your comments. The one I wasn't sure about was the one about invalidating parameter values as non-strings. Can you clarify what you were looking for there?

File: copycat/api/src/test/java/org/apache/kafka/copycat/data/SchemaBuilderTest.java (hunk 2 above)
Comment (guozhangwang) [inline copycat/api/src/test/java/org/apache/kafka/copycat/data/SchemaBuilderTest.java:207]: You are right, I was over-concerned.

Comment (ewencp) [inline copycat/api/src/test/java/org/apache/kafka/copycat/data/SchemaBuilderTest.java:207]: Appropriately concerned. I started writing a test before I realized I couldn't get a non-String value in there :)

**Pull request closed without merging**